From 9948219ba2b1a228028e31f634098f4b5e8add7b Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Sat, 16 May 2026 16:51:17 -0500
Subject: [PATCH 119/121] winewayland: Overwrite previous toplevel contents
 with ensure contents.

In the future when we need to sync the toplevel and client surface contents, we only will sync those when the app tries to draw to the toplevel which will lead to client surface detach followed by a copy.
---
 dlls/winewayland.drv/wayland_surface.c | 11 +++++++----
 dlls/winewayland.drv/waylanddrv.h      |  2 +-
 dlls/winewayland.drv/window.c          |  2 +-
 3 files changed, 9 insertions(+), 6 deletions(-)

diff --git a/dlls/winewayland.drv/wayland_surface.c b/dlls/winewayland.drv/wayland_surface.c
index 5bb30deae5b..ae0eabc33c7 100644
--- a/dlls/winewayland.drv/wayland_surface.c
+++ b/dlls/winewayland.drv/wayland_surface.c
@@ -634,6 +634,7 @@ void wayland_surface_clear_role(struct wayland_surface *surface)
 
     surface->content_width = 0;
     surface->content_height = 0;
+    surface->needs_contents = FALSE;
 
     wl_display_flush(process_wayland.wl_display);
 }
@@ -1576,14 +1577,15 @@ void wayland_surface_ensure_contents(struct wayland_surface *surface)
     width = surface->window.rect.right - surface->window.rect.left;
     height = surface->window.rect.bottom - surface->window.rect.top;
     needs_contents = surface->window.visible &&
-                     (surface->content_width != width ||
-                      surface->content_height != height);
+                     ((surface->content_width != width ||
+                       surface->content_height != height) ||
+                       surface->needs_contents);
+
+    if (!needs_contents) return;
 
     TRACE("surface=%p hwnd=%p needs_contents=%d\n",
           surface, surface->hwnd, needs_contents);
 
-    if (!needs_contents) return;
-
     if (wayland_surface_reconfigure(surface))
     {
         /* Create a transparent dummy buffer. */
@@ -1601,6 +1603,7 @@ void wayland_surface_ensure_contents(struct wayland_surface *surface)
 
         wayland_surface_attach_shm(surface, dummy_shm_buffer, damage);
         wl_surface_commit(surface->wl_surface);
+        surface->needs_contents = FALSE;
     }
 
     if (damage) NtGdiDeleteObjectApp(damage);
diff --git a/dlls/winewayland.drv/waylanddrv.h b/dlls/winewayland.drv/waylanddrv.h
index ff2936410fd..d0ea92d1aef 100644
--- a/dlls/winewayland.drv/waylanddrv.h
+++ b/dlls/winewayland.drv/waylanddrv.h
@@ -390,7 +390,7 @@ struct wayland_surface
     };
 
     struct wayland_surface_config pending, requested, processing, current;
-    BOOL resizing;
+    BOOL resizing, needs_contents;
     struct wayland_window_config window;
     int content_width, content_height;
     HCURSOR hcursor;
diff --git a/dlls/winewayland.drv/window.c b/dlls/winewayland.drv/window.c
index 890dcddbe49..573dbc7e8e4 100644
--- a/dlls/winewayland.drv/window.c
+++ b/dlls/winewayland.drv/window.c
@@ -995,7 +995,7 @@ BOOL set_window_surface_contents(HWND hwnd, struct wayland_shm_buffer *shm_buffe
         {
             wayland_surface_attach_shm(wayland_surface, shm_buffer, damage_region);
             wl_surface_commit(wayland_surface->wl_surface);
-            committed = TRUE;
+            wayland_surface->needs_contents = committed = TRUE;
         }
         else
         {
-- 
2.54.0

